[Routing] Add seamless support for unicode requirements

nicolas-grekas opened the pull request:
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | #3629, #5236, #19562 |
| License | MIT |
| Doc PR | https://github.com/symfony/symfony-docs/pull/6890 |

This PR adds unicode support to route matching and generation by automatically adding the `u` modifier to regexps that use either unicode characters or unicode enabled character classes (e.g. `\p...` `\x{...}` `\X`).

As a side note, if one wants to match a single unicode character (vs a single byte), one should use `\PM` or `\X` instead of `.` _or_ set the `unicode` parameter to true.

nicolas-grekas committed:
[Routing] Add seamless support for unicode requirements

nicolas-grekas commented:
As a side, if one wants to match a single unicode character (vs a single byte), one should use `\PM` instead of `.`.

nicolas-grekas commented:
ping @Tobion

c960657 commented:
I like how this approach is simpler than my own PR #19562. I think UTF-8 is the only reasonable encoding to use for URLs (because this is what browsers use when pretty-printing the URLs in the location field), and I assume most people will use UTF-8 internally in their data and source code, so I like how this patch favours UTF-8 while still allowing people to do otherwise if they want.

I think the auto-detection of UTF-8 patterns is a bit too magic. E.g. whether `.` matches a byte or a character is determined by the whether the pattern contains some of the triggering characters elsewhere. So I am wondering whether we can make a more explicit but backwards-compatible way of enabling/disabling UTF-8.

Here are two examples that are explicit. They are not native PCRE syntax but reuse syntax used in PCRE.

 1.  PCRE allows enabling UTF-8 mode from within the pattern using `(*UTF8)` (mentioned [here](http://www.pcre.org/pcre.txt). This is only valid at the beginning of the pattern, but we could reuse the syntax for triggering UTF-8 and just strip it when concatenating the pattern.
 2.  PCRE allows setting [internal options](http://php.net/manual/en/regexp.reference.internal-options.php) using `(?`, e.g. `(?i)` and `(?-i)`will enable and disable case-insensitivity, respectively. This does not work for the `u` pattern, but perhaps we could reuse the syntax and strip it. This syntax also allows us to make UTF-8 on by default in Symfony 4, so that it has to be explicitly disabled.

According to the [manual](http://php.net/manual/en/regexp.reference.unicode.php), using `\w` is faster than `\pL`, so it would be nice if one could trigger UTF-8 mode without using `\pL`.

nicolas-grekas commented:
I didn't know about the `(*UTF8)` prefix thanks for the link! I don't think we should target adding the `u` flag by default in 4.0: most of the time, URLs are plain ASCII string. But I agree with you: having magic behavior for `.` is an issue and recommending `\X` or `\PM` to opt into unicode comes with a perf overhead.
Since unicode should really be enabled at the route level (not at the requirement level), I thing we could add a conventional prefix to routes for enabling unicode. I propose `*` for now.
The current magic that detects unicode could be turned into a deprecation (an exception in 4.0) for warning the user when one uses unicode characters/properties while the `*` prefix is missing.

nicolas-grekas commented:
The prefix idea doesn't play well with route prefixing as done by loaders...
The PR now handles a `*` prefix to enable unicode at the requirements level.
See test cases also.

stof commented:
Another solution is to use a route option to opt in unicode mode. Route options are precisely meant to give hints to the route compiler.
This would be easier to explain and remember than the fact of adding a magical ``(*UTF8)`` needing to be stripped.

nicolas-grekas commented:
@stof but that would be a DX nightmare: if you'd use unicode, you'd need to repeat yourself thousands of times. The current way is a bit magic, but really seamless: if you use any unicode chars or any unicode property, unicode is enabled. There is only one special case: the `*` prefix to requirements to force unicode matching for `.` when needed.

nicolas-grekas commented:
Doc PR added: https://github.com/symfony/symfony-docs/pull/6890

fabpot commented:
I don't really like the `*` convention, but I don't have any better idea, so 👍

stof commented:
@fabpot what about my proposal of using my proposal ?

@nicolas-grekas we could keep the autodetection of unicode. The option would be used instead of using a magic ``*`` (the other question is whether setting the option to false explicitly should disable the autodetection)

fabpot commented:
@stof Indeed, I think an option is better than an obscure convention.

nicolas-grekas commented:
updated

c960657 commented:
The patch uses the term “Unicode” in variable names etc., but wouldn't it be more correct to use “UTF-8” (the specific encoding we are supporting)? “Unicode” is a much wider concept. Note that the Unicode character properties escape sequences, `\p{xx}` etc., works even in non-UTF-8-mode.

Also, I think we disagree a bit on what exactly this feature does:
```
    /**
     * Returns the unicode enforcement status.
     *
     * @return bool Whether unicode matching is enforced or not
     */
    public function getUnicode()
```
In my eyes, this flag enabled UTF-8 support for regular expressions, i.e. basically adds the `u` modifier – nothing else. Do you agree, or could you elaborate what "unicode enforcement"?

c960657 commented on src/Symfony/Component/Routing/CHANGELOG.md:
```diff
@@ -5,6 +5,7 @@ CHANGELOG
 
  * Added support for `bool`, `int`, `float`, `string`, `list` and `map` defaults in XML configurations.
+ * Added support for unicode requirements
   
```
Unicode is a proper noun, so it should be capitalized.

c960657 commented on src/Symfony/Component/Routing/Route.php:
```diff
@@ -65,6 +65,11 @@ class Route implements \Serializable
+     * @var bool
+     */
+    private $unicode = '';
+
```
The variable is a boolean, so the default should be `false`, not an empty string.

c960657 commented on src/Symfony/Component/Routing/RouteCollection.php:
```diff
@@ -200,6 +200,20 @@ public function addDefaults(array $defaults)
+     * Existing settings will be overridden.
+     *
+     * @param string $unicode Whether unicode matching is enforced or not
+     */
```
@param bool

c960657 commented on src/Symfony/Component/Routing/RouteCompiler.php:
```diff
@@ -91,6 +91,12 @@ private static function compilePattern(Route $route, $pattern, $isHost)
         $pos = 0;
         $defaultSeparator = $isHost ? '.' : '/';
+        $useUnicode = preg_match('//u', $pattern);
+        $needsUnicode = $route->getUnicode() || ($useUnicode && preg_match('/[\x80-\xFF]/', $pattern));
```
I don't understand the purpose of this detection logic. Why not just rely on `$route->getUnicode()`?

nicolas-grekas replied:
That's why I added "seamless" to the description. From the doc PR:

    Since Symfony 3.2, requirements that use any non-ASCII UTF-8 characters or a
    a `PCRE Unicode property`_ (``\p{xx}``, ``\P{xx}`` or ``\X``) turn on the
    `unicode` parameter on its corresponding route. You can also set this
    `unicode` parameter to true explicitly when declaring or importing routes so
    that e.g. a `.` matches any UTF-8 characters instead of just a single byte. 

This required to reduce the boilerplate. People want to write `[éè]` and have it work out of the box.

c960657 replied:
While this is seamless most of the time, it leaves some surprising edge cases that cannot easily be fixed. The pattern `[éè]` works out of the box, so does `[éè].`, but `[ef].` suddenly requires you to explicitly enable UTF-8 matching. Of course, this can all be explained in the documentation. I just think its a bit confusing, especially because it is a bit unclear why some escape sequences have been chosen to trigger Unicode mode and others have not.

Well, this is just my gut feeling about this. Perhaps others find it easier to understand.

nicolas-grekas replied:
> but `[ef].` suddenly requires you to explicitly enable UTF-8 matching

at least for bc reasons yes, this can't be changed easilly

c960657 commented on src/Symfony/Component/Routing/RouteCompiler.php:
```diff
@@ -140,6 +154,15 @@ private static function compilePattern(Route $route, $pattern, $isHost)
+            } else {
+                if (!preg_match('//u', $regexp)) {
+                    $useUnicode = false;
+                } elseif (!$needsUnicode && preg_match('/[\x80-\xFF]|(?<!\\\\)\\\\(?:\\\\\\\\)*+(?-i:X|[pP][\{CLMNPSZ]|x\{)/', $regexp)) {
```
This seems magic. Do we need to override `$route->getUnicode()` here?

nicolas-grekas replied:
See above.

c960657 replied:
This pattern should match both `\x{` and just `\x`. For character codes < 256, \xB3 and \x{B3} are equivalent.

nicolas-grekas replied:
> \xB3 and \x{B3} are equivalent

they are, but as you know, `/\xB3/u` is not equivalent to `/\xB3/` (multibyte utf-8 vs single byte high-ASCII). This is an edge case. In order to not falsly turn on unicode, this makes me think we should auto-enable only when using 3 or more hexadecimal numbers between brackets.

nicolas-grekas replied:
updated

nicolas-grekas commented:
> The patch uses the term “Unicode” in variable names etc., but wouldn't it be more correct to use “UTF-8” 

I hesitated on this, and choose Unicode to match what the PCRE doc uses. On the web, unicode is synonym for UTF-8. Nobody uses any other Unicode encoding there. And the PCRE doc says about "unicode properties", the "unicode" flag, etc. I thought it'd be better to make the vocabularies match.

> Unicode character properties escape sequences, \p{xx} etc., works even in non-UTF-8-mode.

Yes, it's undocumented (in the PHP doc at least, where on the contrary it's specifically documented under the "Unicode properties" chapter), and thus nobody knows what it does, esp. when considering high-ASCII chars.

> this flag [...] adds the u modifier – nothing else.

Yes when given `true`. But when given `false`, it doesn't "remove" the `u` modifier. Instead, it then relies on detecting if a unicode char or prop to also add the `u` modifier.
Thus this flag "enforces" the `u` modifier, whereas setting this flag to false "turn on" auto-detection.

This means that 

 Do you agree, or could you elaborate what "unicode enforcement" means?

c960657 commented:
>Yes, yet it's undocumented (in the PHP doc at least, where on the contrary it's specifically documented under the "Unicode properties" chapter), and thus nobody knows what it does, esp. when considering high-ASCII chars.

On the [man page](http://linux.die.net/man/3/pcrepattern) it says:
"When in 8-bit non-UTF-8 mode, these sequences are of course limited to testing characters whose codepoints are less than 256, but they do work in this mode."

nicolas-grekas commented:
I never read the man page of pcre but I read many times the one on php.net. I guess I'm more like the usual PHP user on this one :)

c960657 commented:
>I never read the man page of pcre but I read many times the one on php.net.

I'm not sure which parts of the PHP manual you are referring to. I couldn't find any occurrences of the phrase "Unicode mode" outside user comments. "UTF-8 mode" occurs a few places, e.g. in the [description of PREG_BAD_UTF8_OFFSET_ERROR](http://php.net/manual/en/pcre.constants.php).

Unicode mode:
https://www.google.dk/search?q=%22unicode+mode%22+site%3Aphp.net+inurl%3Amanual

UTF-8 mode:
https://www.google.dk/search?q=%22utf-8+mode%22+site%3Aphp.net+inurl%3Amanual

Additionally, the work "Unicode" is used in connection with character classes, but as mentioned above they work with and without UTF-8 mode.

nicolas-grekas commented:
@c960657 thanks for you input. I've just renamed unicode to utf8 everywhere.

I've also added a deprecation targeting the current magic: I propose to encourage people to explicitly enable the utf8 flag; throw a deprecation when they don't, and throw a LogicException in 4.0.

c960657 commented:
Sounds like a good solution :)

stof commented on src/Symfony/Component/Routing/Annotation/Route.php:
```diff
@@ -143,4 +144,14 @@ public function getCondition()
+        $this->utf8 = $utf8;
+    }
+
+    public function isUtf8()
```
do we need a new property on the route ? I suggested using options, which are already meant to provide hints to the route compiler

nicolas-grekas replied:
That would complicate the transition: options are validated and you can't add an unsupported key (no forward compat). This means it wouldn't be possible to write routes in a forward/backward compatible manner in Yml/Xml/Annotations.

nicolas-grekas commented on src/Symfony/Component/Routing/Route.php:
```diff
@@ -70,6 +70,7 @@ class Route implements \Serializable
      * Available options:
      *
      *  * compiler_class: A class name able to compile this route instance (RouteCompiler by default)
+     *  * utf8:           Whether UTF-8 matching is enforced ot not
```
utf8 is now an option on the route. ping @stof (I don't know were I've got this idea that it wasn't possible...)

stof commented on src/Symfony/Component/Routing/RouteCompiler.php:
```diff
@@ -168,10 +196,21 @@ private static function compilePattern(Route $route, $pattern, $isHost)
+        // enable Utf8 matching if really required
+        if ($needsUtf8) {
+            $regexp .= 'u';
+            for ($i = count($tokens) - 1; $i >= 0; --$i) {
```
why not just iterating over tokens ? (we don't actually need the reverse ordering)

nicolas-grekas replied:
because that's exactly how it's done a few lines above...

nicolas-grekas replied:
order reversed

stof commented:
:+1:

fabpot commented:
Thank you @nicolas-grekas.
